olive/passes/onnx/float32_conversion.py

```diff
@@ -0,0 +1,126 @@
+# -------------------------------------------------------------------------
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License.
+# --------------------------------------------------------------------------
+import logging
+import re
+from collections import defaultdict
+from pathlib import Path
+from typing import Any, Dict
+
+import onnx
+
+from olive.hardware.accelerator import AcceleratorSpec
+from olive.model import ONNXModelHandler
+from olive.model.utils import resolve_onnx_path
+from olive.passes import Pass
+from olive.passes.onnx.common import get_external_data_config, model_proto_to_olive_model
+from olive.passes.pass_config import PassConfigParam
+
+logger = logging.getLogger(__name__)
+
+
+class OnnxIOFloat16ToFloat32(Pass):
+    """Converts float16 model inputs/outputs to float32."""
+
+    @classmethod
+    def _default_config(cls, accelerator_spec: AcceleratorSpec) -> Dict[str, PassConfigParam]:
+        config = {
+            "name_pattern": PassConfigParam(
+                type_=str,
+                default_value="logits",
+                description=(
+                    "Only convert inputs/outputs whose name matches this pattern. By defaultlooking for logits names"
+                ),
+            )
+        }
+        config.update(get_external_data_config())
+        return config
+
+    def create_io_mapping(self, graph, i_map, o_map):
```

Comment:
`OnnxDAG` doesn't cover this use case fully but it can be extended to allow us this type of graph manipulation if needed. The logic here is okay for now but we could update this in the future to use onnx dag. it has an easily accessible graph representation so that we don't have to loop through the graph to make changes.

Reply:
added this pass to workitem 31579 tracking onnx dag use cases.